Add `webgl_loader_multiple.html` example with drag-and-drop loading

hybridherbst opened the pull request:
**Description**

This PR adds drag-and-drop file loading to the basic webgl_loader_gltf example.
As this demonstrates how to build a basic file viewer with three.js, another option would be to make a separate example file (e.g. webgl_file_viewer).

Key questions for this PR would thus be:
- would this be a good fit for the three.js examples in general
- would a "basic viewer" be a good separate example instead of putting it into the loader_gltf sample.

This contribution is funded by [needle](https://needle.tools).

hybridherbst committed:
add webgl_loader_multiple example that supports glTF and FBX loading, proper disposal, and animation

* added webgl_loader_multiple as example of how to load multiple files one after each other
* added bounds fitting
* added FBX import as second file type example
* added material upgrade of FBX materials to MeshStandardMaterial to receive IBL
* fix minDistance possibly being greater than maxDistance, fix animation loading and playback
* switched to animationLoop

workaround for having a nicely blurred background from the environment

alternative blur approach: use texture mipmaps directly. less pretty

clean up, add link to PR that will introduce background bluriness

fix env bluriness (harder with mipmaps), fix HDRI Haven url

update screenshot

replace blurred background hack with new backgroundBlurriness

feat: add USDZ loading, glTF export and USDZ export, FBX material conversion

USDZExporter: fix default plane anchoring orientation and add precision specifiers (#25012)

Fixes #25011

[XR] Feat: Allow passing GLTFLoader into OculusHandModel/XRHandMeshModel (#25013)

* feat: allow passing GLTFLoader into OculusHandModel/XRHandMeshModel

* apply review feedback

USDZExporter: guard against missing image data in textures

USDZLoader: add hierarchy support and fix loading files exported with USDZExporter

GLTFExporter: guard against missing image data in textures

simplify null check

cleanup child name and add example for texture traversal after loading

added file input button, added download button for OBJ

fix: GLTF and USDZ exporters break with missing images in textures

adjust title and description, convert OBJ mats to MeshStandardMaterial so they use IBL

updated screenshot

WestLangley commented:
You need to properly `dispose()` of the the old model.

//

Also, compare your implementation of drag and drop with the implementation in the Matcap example, which itself is based on code from the Editor.

(I am not saying one implementation is better than the other, but maintaining consistent patterns is preferable.)

takahirox commented:
> another option would be to make a separate example file (e.g. webgl_file_viewer).

Personally I prefer to make a separate example if we want a viewer example. glTF is a standard 3D format and webgl_loader_gltf example may be one of the best example for new users to start Three.js. It would be good to keep it as simple as possible.

But as the diff shows tha change isn't so big. Others might think it would be acceptable tho.

Mugen87 commented:
I vote to create a separate example and keep `webgl_loader_gltf` as simple as possible.

hybridherbst commented:
That's great too, as it means it can get some other example viewer features over time :)

Would webgl_viewer_gltf work? Or just webgl_viewer? It would actually be cool to have a multi-file example.

@WestLangley I'll look at that implementation, thanks!

Mugen87 commented:
The thing is the `three.js` editor is already some kind of viewer. I'd like to avoid that we implement the same thing multiple times, see:

https://github.com/mrdoob/three.js/blob/55f98507d510fdf3e8c374982c5ded09d7aa7239/editor/js/Loader.js#L262-L285

Hence, I'm hesitate a bit about this PR...

hybridherbst commented:
The three.js editor isn't a good file viewer, it's a basic editor. 
For example:
- Dropping a file results in a black view (needs to manually add a light). 
- Dropping another file superimoses that (need to manually delete the previous file).
- Not a good example for how to build a viewer (too complex)



For context, I made this PR because to simply look at a glTF file on the dev branch (to check https://github.com/mrdoob/three.js/issues/23764) there's a cumbersome number of steps needed, as the editor comes with its own set of issues, bugs, and complexity. The changes above were actually what I did just to test a simple glTF file, and then thought that it would make sense if that worked by default.

Mugen87 commented:
Fair point that the current editor is not 100%  comparable to a pure viewer. However there is also https://github.com/donmccurdy/three-gltf-viewer. Wouldn't we essentially implement the same thing in the examples now?

hybridherbst commented:
I think that's also a difference between a "complete implementation" and "providing a good starting example" – Don's viewer is great but not a "minimal example" (as I understand three's examples). For me, it was mostly "how about it's faster to test files"; same as for the MatCap viewer the drag-drop functionality isn't exactly required but serves as example on how to use that part of three.

Regarding freeing resources: @WestLangley mentioned that the resources should be disposed but @Mugen87 replied here https://discourse.threejs.org/t/how-to-dispose-and-destroy-gltf-object-completely/24761/2 that just removing from the scene should be sufficient. Which one is it? :) And if they are to be disposed: would traversing the root and disposing all meshes/materials/textures be sufficient?

Mugen87 commented:
>  @Mugen87 replied here https://discourse.threejs.org/t/how-to-dispose-and-destroy-gltf-object-completely/24761/2 that just removing from the scene should be sufficient.

I'm afraid you misunderstood my post^^. The OP at discourse removed the object twice. That's why I said one call of `remove()` is sufficient. Using `dispose()` is mandatory.

Mugen87 commented:
The problem is the definition "minimal example". What does that entail? Eventually, we can quickly end up with all basic features required for a PBR viewer. Meaning drag'n'drop, IBL setup, animation playback, automatic camera adjustment, support from compressed assets (KTX/DRACO) and so on. This gets quickly too much. Hence, I vote to not add a viewer example. Don's viewer is compact enough as a template for a basic viewer implementation.

hybridherbst commented:
Thanks for the clarification! I kinda disagree that a 750 line file is a good basic example if all that's needed to bring "here's a single model" to a basic "you can drop your own" viewer are like 10 lines. If you can find your way around in that that's great but then you don't need an example anymore anyways.

Follow-up question would be, already a good case for "why would a file loading example in three be good to have": what's the right way to track the meshes/materials/textures loaded via the file? 
- Traverse meshes/materials, iterate all material keys and check if they are textures?
- I looked at LoadingManager, should one manually keep track of assets that were loaded there?

The [implementation in Don's viewer](https://github.com/donmccurdy/three-gltf-viewer/blob/2784c12c4917eda6a97e0b101951a5a47c066bc6/src/viewer.js#L704) has a hardcoded list of textures to dispose which will for example not track if textures are loaded as part of extensions (e.g. volume/transmission textures would already not be tracked there).

WestLangley commented:
For glTF format, don't you need to support dropping of folders?

Also, you would need to reset the camera properties and the control distance limits to accommodate the bounding box of the new object.

I'm inclined to agree that it is better to keep this example simple.

hybridherbst commented:
I'd be fine with supporting only .glb (haven't spent much testing on .gltf files) for the purpose of a dedicated sample.
Regarding the scope, to me both of these points (how do I calculate nice bounds to frame a model, how to properly clean up resources) HAVE come up in production needs and then there was no example for them, making solving these challenges unnecessarily hard. 

Past me would certainly have been happy about a dedicated, simple gltf viewer example :)

hybridherbst commented:
Any further input here?
I've updated the original PR with the comments/questions above (proper disposal of data, following the same pattern as WestLangley suggested)

I could also turn this into a minimal example that would cover the following topics in exemplary fashion:
- receive drag-drop input for GLTFLoader
- dispose of already created data
- calculates frame bounds when dropping a model.

Mugen87 commented:
In any case, I don't think `webgl_loader_gltf` should be made more complex. Consider to create a new example `webgl_viewer_gltf`.

In the next step, @mrdoob has to decide whether a viewer example should be part of the repo or not. Depending on his decision, further input might come up.

hybridherbst commented:
I moved the code to a new `webgl_loader_multiple.html`.

It serves as an example for how to:
- load files on drag and drop (two types, FBX and GLB, each with embedded textures only to keep the example simple)
- unload and dispose objects, materials, textures
- calculate approximate scene bounds and fit camera controls

WestLangley commented:
The demo is freezing up on me. It appears to happen when `.maxDistance` is computed to be less than `.minDistance`.

hybridherbst commented:
@WestLangley do you have a repro model file?

takahirox commented:
Any special reasons why you want to support not only glb but also fbx? I thought it would be good to keep the example code as simple as possible so prefer to support only glb.

takahirox commented on examples/webgl_loader_multiple.html:
```diff
@@ -0,0 +1,314 @@
+<!DOCTYPE html>
+<html lang="en">
+	<head>
+		<title>three.js webgl - GLTFloader</title>
```
Maybe title should be clearer. And if we support not only glb but also fbx, this title mismatches.

hybridherbst replied:
Adjusted to three.js webgl - Multi-File Viewer (FBX and glTF)

takahirox commented on examples/webgl_loader_multiple.html:
```diff
@@ -0,0 +1,314 @@
+						loader.setKTX2Loader( ktx2Loader );
+						loader.setMeshoptDecoder( MeshoptDecoder );
+
+						loader.load( 'DamagedHelmet.gltf', gltf => addToScene( gltf.scene ) );
```
Personally, I think an empty scene may be ok by default. And users might be a bit confused if they see two damaged helmet+Royal Esplanade background screenshots in the example list.

hybridherbst replied:
Removed the default file, good point

takahirox commented on examples/webgl_loader_multiple.html:
```diff
@@ -0,0 +1,314 @@
+						const loader = new GLTFLoader();
+						loader.setKTX2Loader( ktx2Loader );
+						loader.setMeshoptDecoder( MeshoptDecoder );
+						loader.parse( reader.result, '', gltf => addToScene( gltf.scene ) );
```
To clarify, you don't think we need to support animations in this example, correct?

hybridherbst replied:
I'd prefer if animations would be supported – I see it as an example to consolidate ideas around "loading files and viewing them" into one place. Added that again

hybridherbst commented:
I've fixed the title, adjusted the HDRI, adjusted min-max logic, and refactored animation playback to ensure it works for both glTF and FBX.
The reason for adding both was to have an "loader_multiple" example that shows how to deal with the differences and similarities in loaders and scene structures.

I've also added a background bluriness placeholder based on https://github.com/mrdoob/three.js/pull/23712; in the commit history is also a version that uses PMREM instead of simple mipmaps but that's a different discussion.

Having this example has already made working on three.js a lot easier for me, as there's now a nice basic example to add more advanced viewer features on top, or to test out glTF extensions, post effects on different models, ...

hybridherbst commented:
Rebased on dev. I still find this incredibly useful to 
a) have a reference for implementing drag-drop loaders
b) a quick way to check what three supports / doesn't support when it comes to glTF (and FBX and potentially other formats).

hybridherbst commented:
Rebased on latest dev, and replaced the background blurriness workaround with `backgroundBlurriness`.

hybridherbst commented:
Rebased to latest three. 

I can also close this PR if there's no interest in a more complete "here's how to build a viewer" sample.
It has been super valuable for me to have this example while getting started with three.js and contributing to it.
I keep referencing it a lot, and test most of PRs related to Loaders/Exporters in three.js with this sample.

mrdoob commented:
Hmm, are the `GLTFExporter` and `USDZExporter` changes in the PR needed?

hybridherbst commented:
Nope, they're not, accidentally wrong rebase order.

As there wasn't a response here I assumed there's no interest and probably added some things not necessarily wanted for a PR. I can clean that up – or close this PR – your choice!

hybridherbst commented:
Small correction: they were/are necessary related to 
- https://github.com/mrdoob/three.js/pull/25022

which was decided to not be the right approach. I found files that resulted in invalid textures when loaded via FBXLoader, nobody showed interested in me even attempting to fix FBXLoader, so I wanted to guard against the data on export, which was also not desired. I think it's rare enough...

Mugen87 commented:
Closing. The proposed `webgl_loader_multiple` is really more an application showcase and not an example that demonstrates a single feature or component of the engine. Ideally, `webgl_loader_multiple` is located in a separate repository and shared in the Showcase category of the forum.
